In [43]:
#Imports 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


# Data Exploration

In [44]:
df_order=pd.read_csv("C:/Users/jwsam/Desktop/Ecom/kaggle_datasets/olist_orders_dataset.csv")

In [45]:
df_order.head(60)

order_id                       customer_id  \
0   e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1   53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2   47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3   949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4   ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5   a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6   136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7   6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8   76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9   e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   
10  e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
11  34513ce0c4fab462a55830c0989c7edb  7711cf624183d843aafe81855097bc37   
12  82566a660a982b15fb86e904c8d32918  d3e3b74c766bc6214e0c830b17ee2341   
13  5ff96c15d0b717ac6ad1f3d77225a350  19402a48fe860416adf93348aba37740   
14  432aaf21d85167c2c86ec9448c4e42cc  3df704f53d3f1d4818840b34ec672a9f   
15  dcb36b511fcac050b97cd5c05de84dc3  3b6828a50ffe546942b7a473d70ac0fc   
16  403b97836b0c04a622354cf531062e5f  738b086814c6fcc74b8cc583f8516ee3   
17  116f0b09343b49556bbad5f35bee0cdf  3187789bec990987628d7a9beb4dd6ac   
18  85ce859fd6dc634de8d2f1e290444043  059f7fc5719c7da6cbafe370971a8d70   
19  83018ec114eee8641c97e08f7b4e926f  7f8c8b9c2ae27bf3300f670c3d478be8   
20  203096f03d82e0dffbc41ebc2e2bcfb7  d2b091571da224a1b36412c18bc3bbfe   
21  f848643eec1d69395095eb3840d2051e  4fa1cd166fa598be6de80fa84eaade43   
22  2807d0e504d6d4894d41672727bc139f  72ae281627a6102d9b3718528b420f8a   
23  95266dbfb7e20354baba07964dac78d5  a166da34890074091a942054b36e4265   
24  f3e7c359154d965827355f39d6b1fdac  62b423aab58096ca514ba6aa06be2f98   
25  fbf9ac61453ac646ce8ad9783d7d0af6  3a874b4d4c4b6543206ff5d89287f0c3   
26  acce194856392f074dbf9dada14d8d82  7e20bf5ca92da68200643bda76c504c6   
27  dd78f560c270f1909639c11b925620ea  8b212b9525f9e74e85e37ed6df37693e   
28  91b2a010e1e45e6ba3d133fa997597be  cce89a605105b148387c52e286ac8335   
29  ecab90c9933c58908d3d6add7c6f5ae3  761df82feda9778854c6dafdaeb567e4   
30  f70a0aff17df5a6cdd9a7196128bd354  456dc10730fbdba34615447ea195d643   
31  1790eea0b567cf50911c057cf20f90f9  52142aa69d8d0e1247ab0cada0f76023   
32  989225ba6d0ebd5873335f7e01de2ae7  816f8653d5361cbf94e58c33f2502a5c   
33  d887b52c6516beb39e8cd44a5f8b60f7  d9ef95f98d8da3b492bb8c0447910498   
34  b276e4f8c0fb86bd82fce576f21713e0  cf8ffeddf027932e51e4eae73b384059   
35  8563039e855156e48fccee4d611a3196  5f16605299d698660e0606f7eae2d2f9   
36  60550084e6b4c0cb89a87df1f3e5ebd9  f5458ddc3545711efa883dd7ae7c4497   
37  5acce57f8d9dfd55fa48e212a641a69d  295ae9b35379e077273387ff64354b6f   
38  434d158e96bdd6972ad6e6d73ddcfd22  2a1dfb647f32f4390e7b857c67458536   
39  7206b86ea789983f7a273ea7fa0bc2a8  3391c4bc11a817e7973e498b0b023158   
40  1e7aff52cdbb2451ace09d0f848c3699  ddaff536587109b89777e0353215e150   
41  6ea2f835b4556291ffdc53fa0b3b95e8  c7340080e394356141681bd4c9b8fe31   
42  948097deef559c742e7ce321e5e58919  8644be24d48806bc3a88fd59fb47ceb1   
43  d22e9fa5731b9e30e8b27afcdc2f8563  756fb9391752dad934e0fe3733378e57   
44  ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
45  6ebaec694d7025e2ad4a05dba887c032  4f28355e5c17a4a42d3ce2439a1d4501   
46  d17dc4a904426827ca80f2ccb3a6be56  569cf68214806a39acc0f39344aea67f   
47  25f4376934e13d3508486352e11a5db0  12fd2740039676063a874b9567dfa651   
48  5820a1100976432c7968a52da59e9364  2b56e94c2f66f2d97cfa63356f69cee8   
49  2ce1ad82022c1ba30c2079502ac725aa  7f2178c5d771e17f507d3c1637339298   
50  138849fd84dff2fb4ca70a0a34c4aa1c  9b18f3fc296990b97854e351334a32f6   
51  47aa4816b27ba60ec948cd019cc1afc1  148348ff65384b4249b762579532e248   
52  9faeb9b2746b9d7526aef5acb08e2aa0  79183cd650e2bb0d475b0067d45946ac   
53  641fb0752bf5b5940c376b3a8bb9

In [46]:
df_order.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


### ALl the features share same number of instances: 99,441 
except for 
-  order_approved_at:             99281
- order_delivered_carrier_date:  97658  
- order_delivered_customer_date: 96476 

In [47]:
df_order.describe()

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-03-31 15:08:21  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-14 20:02:44   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

In [48]:
df_order['order_approved_at'].isna().sum()

np.int64(160)

In [49]:
df_order['order_status'].unique()

<StringArray>
[  'delivered',    'invoiced',     'shipped',  'processing', 'unavailable',
    'canceled',     'created',    'approved']
Length: 8, dtype: str

In [50]:
df_order[df_order['order_status']=='unavailable']

order_id                       customer_id  \
266    8e24261a7e58791d10cb1bf9da94df5c  64a254d30eed42cd0e6c36dddb88adf0   
586    c272bcd21c287498b4883c7512019702  9582c5bbecc65eb568e2c1d839b5cba1   
687    37553832a3a89c9b2db59701c357ca67  7607cd563696c27ede287e515812d528   
737    d57e15fb07fd180f06ab3926b39edcd2  470b93b3f1cde85550fc74cd3a476c78   
1160   2f634e2cebf8c0283e7ef0989f77d217  7353b0fb8e8d9675e3a704c60ca44ebe   
...                                 ...                               ...   
99147  6cf5c427fb11bc5fe8abefa86aa7ba0b  91ac14a1e50bc022eeecf0c2ba37006a   
99233  1c7c8ab324927f5b6e52338bd110bef4  95b1767df51d6fa09402267d8de314df   
99252  aaab15da689073f8f9aa978a390a69d1  df20748206e4b865b2f14a5eabbfcf34   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp    order_approved_at  \
266    unavailable      2017-11-16 15:09:28  2017-11-16 15:26:57   
586    unavailable      2018-01-31 11:31:37  2018-01-31 14:23:50   
687    unavailable      2017-08-14 17:38:02  2017-08-17 00:15:18   
737    unavailable      2018-01-08 19:39:03  2018-01-09 07:26:08   
1160   unavailable      2017-09-27 20:55:33  2017-09-28 01:32:50   
...            ...                      ...                  ...   
99147  unavailable      2017-10-17 10:53:48  2017-10-19 03:12:43   
99233  unavailable      2017-12-23 18:48:09  2017-12-23 18:54:22   
99252  unavailable      2018-01-16 14:27:59  2018-01-17 03:37:34   
99348  unavailable      2017-08-23 16:28:04  2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03  2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
266                            NaN                           NaN   
586                            NaN                           NaN   
687                            NaN                           NaN   
737                            NaN                           NaN   
1160                           NaN                           NaN   
...                            ...                           ...   
99147                          NaN                           NaN   
99233                          NaN                           NaN   
99252                          NaN                           NaN   
99348                          NaN                           NaN   
99415                          NaN                           NaN   

      order_estimated_delivery_date  
266             2017-12-05 00:00:00  
586             2018-02-16 00:00:00  
687             2017-09-05 00:00:00  
737             2018-02-06 00:00:00  
1160            2017-10-27 00:00:00  
...                             ...  
99147           2017-10-27 00:00:00  
99233           2018-01-19 00:00:00  
99252           2018-02-06 00:00:00  
99348           2017-09-15 00:00:00  
99415           2017-10-23 00:00:00  

[609 rows x 8 columns]

In [51]:
df_order[df_order['order_status']=='canceled']

order_id                       customer_id  \
397    1b9ecfe83cdc259250e1a8aca174f0ad  6d6b50b66d79f80827b6d96751528d30   
613    714fb133a6730ab81fa1d3c1b2007291  e3fe72696c4713d64d3c10afe71e75ed   
1058   3a129877493c8189c59c60eb71d97c29  0913cdce793684e52bbfac69d87e91fd   
1130   00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801   ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
...                                 ...                               ...   
98791  b159d0ce7cd881052da94fa165617b05  e0c3bc5ce0836b975d6b2a8ce7bb0e3e   
98909  e49e7ce1471b4693482d40c2bd3ad196  e4e7ab3f449aeb401f0216f86c2104db   
99143  6560fb10610771449cb0463c5ba12199  0d07d0a588caf93cc66b7a8aff86d2fe   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   

      order_status order_purchase_timestamp    order_approved_at  \
397       canceled      2018-08-04 14:29:27  2018-08-07 04:10:26   
613       canceled      2018-01-26 21:34:08  2018-01-26 21:58:39   
1058      canceled      2018-01-25 13:34:24  2018-01-25 13:50:20   
1130      canceled      2018-08-28 15:26:39                  NaN   
1801      canceled      2018-09-20 13:54:16                  NaN   
...            ...                      ...                  ...   
98791     canceled      2017-03-11 19:51:36  2017-03-11 19:51:36   
98909     canceled      2018-08-07 11:16:28                  NaN   
99143     canceled      2017-10-01 22:26:25  2017-10-01 22:35:22   
99283     canceled      2018-08-31 16:13:44                  NaN   
99347     canceled      2018-09-06 18:45:47                  NaN   

      order_delivered_carrier_date order_delivered_customer_date  \
397                            NaN                           NaN   
613            2018-01-29 22:33:25                           NaN   
1058           2018-01-26 21:42:18                           NaN   
1130                           NaN                           NaN   
1801                           NaN                           NaN   
...                            ...                           ...   
98791                          NaN                           NaN   
98909                          NaN                           NaN   
99143                          NaN                           NaN   
99283                          NaN                           NaN   
99347                          NaN                           NaN   

      order_estimated_delivery_date  
397             2018-08-14 00:00:00  
613             2018-02-22 00:00:00  
1058            2018-02-23 00:00:00  
1130            2018-09-12 00:00:00  
1801            2018-10-17 00:00:00  
...                             ...  
98791           2017-03-30 00:00:00  
98909           2018-08-10 00:00:00  
99143           2017-10-27 00:00:00  
99283           2018-10-01 00:00:00  
99347           2018-09-27 00:00:00  

[625 rows x 8 columns]

In [52]:
df_order[df_order['order_status']=='approved']

order_id                       customer_id  \
44897  a2e4c44360b4a57bdff22f3a4630c173  8886130db0ea6e9e70ba0b03d7c0d286   
88457  132f1e724165a07f6362532bfb97486e  b2191912d8ad6eac2e4dc3b6e1459515   

      order_status order_purchase_timestamp    order_approved_at  \
44897     approved      2017-02-06 20:18:17  2017-02-06 20:30:19   
88457     approved      2017-04-25 01:25:34  2017-04-30 20:32:41   

      order_delivered_carrier_date order_delivered_customer_date  \
44897                          NaN                           NaN   
88457                          NaN                           NaN   

      order_estimated_delivery_date  
44897           2017-03-01 00:00:00  
88457           2017-05-22 00:00:00

In [53]:
df_order['order_status'].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [54]:
df_order[(df_order['order_status']=='shipped') & (df_order['order_delivered_customer_date'].isna())]

order_id                       customer_id  \
44     ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
154    6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   
162    36530871a5e80138db53bcfd8a104d90  4dafe3c841d2d6cc8a8b6d25b35704b9   
231    4d630f57194f5aba1a3d12ce23e71cd9  6d491c9fe2f04f6e2af6ec033cd8907c   
299    3b4ad687e7e5190db827e1ae5a8989dd  1a87b8517b7d31373b50396eb15cb445   
...                                 ...                               ...   
98534  a59ef0abffbef8ddaae23600b6ee6604  49e64271fb8eeb19851dc7e850e92204   
98779  dab8a6c6bd6ec448df5b3a6b6cb887bc  394653a10cab83cad40d7e2713f3ab89   
98932  492aed3c33bac22a8e04138319829283  58466c1166c377a56f6b2ae0d93ffbc0   
99113  274a7f7e4f1c17b7434a830e9b8759b1  670af30ca5b8c20878fecdafa5ee01b9   
99181  636cdd02667dc8d76d9296bf20a6890a  c162256b133c76f79181ce61d66545db   

      order_status order_purchase_timestamp    order_approved_at  \
44         shipped      2018-06-04 16:44:48  2018-06-05 04:31:18   
154        shipped      2018-01-10 11:33:07  2018-01-11 02:32:30   
162        shipped      2017-05-09 11:48:37  2017-05-11 11:45:14   
231        shipped      2017-11-17 19:53:21  2017-11-18 19:50:31   
299        shipped      2018-06-28 12:52:15  2018-06-28 13:11:09   
...            ...                      ...                  ...   
98534      shipped      2017-11-06 17:35:49  2017-11-07 06:31:44   
98779      shipped      2018-07-14 10:12:51  2018-07-16 12:30:58   
98932      shipped      2018-07-06 16:26:47  2018-07-06 16:35:10   
99113      shipped      2018-06-23 13:25:15  2018-06-23 13:40:11   
99181      shipped      2018-02-17 14:31:22  2018-02-20 07:11:31   

      order_delivered_carrier_date order_delivered_customer_date  \
44             2018-06-05 14:32:00                           NaN   
154            2018-01-11 19:39:23                           NaN   
162            2017-05-11 13:21:47                           NaN   
231            2017-11-22 17:28:34                           NaN   
299            2018-07-04 15:20:00                           NaN   
...                            ...                           ...   
98534          2017-11-07 18:57:39                           NaN   
98779          2018-07-16 14:34:00                           NaN   
98932          2018-07-10 12:27:00                           NaN   
99113          2018-07-04 13:51:00                           NaN   
99181          2018-02-20 19:18:58                           NaN   

      order_estimated_delivery_date  
44              2018-06-28 00:00:00  
154             2018-02-07 00:00:00  
162             2017-06-08 00:00:00  
231             2017-12-13 00:00:00  
299             2018-08-03 00:00:00  
...                             ...  
98534           2017-11-24 00:00:00  
98779           2018-08-03 00:00:00  
98932           2018-07-26 00:00:00  
99113           2018-07-24 00:00:00  
99181           2018-03-14 00:00:00  

[1107 rows x 8 columns]

order_status that are labeled as processing,unavailable,invoiced,app are considered not sent as not ordered

### Since the order_estimated_delivery_date is not null even when the order_status is unavailable or cancled which means this is generated and not applicable to real-life scenarios it is a theoritical estimation based on the order input time

To Further understand consumer behavior towards buying behavior and delivery satisfaction
The data need to be categorized to match the condition that the order was set to be

- Successfully Delivered	
- In Transit	
- Canceled Before Delivery	
- Canceled After Shipment	
- Never Fulfilled

In [55]:
df_order.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

In [56]:
df_order['order_purchase_timestamp'] = pd.to_datetime(df_order['order_purchase_timestamp'])
df_order['order_approved_at'] = pd.to_datetime(df_order['order_approved_at'])
df_order['order_delivered_carrier_date'] = pd.to_datetime(df_order['order_delivered_carrier_date'])
df_order['order_delivered_customer_date'] = pd.to_datetime(df_order['order_delivered_customer_date'])

In [57]:
df_order

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[99441 rows x 8 columns]

In [58]:
df_order['cancel_delay_days'] = (df_order['order_delivered_carrier_date'] - df_order['order_purchase_timestamp']).dt.days

In [59]:
df_order

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  cancel_delay_days  
0               2017-10-18 00:00:00                2.0  
1               2018-08-13 00:00:00                1.0  
2               2018-09-04 00:00:00                0.0  
3               2017-12-15 00:00:00                3.0  
4               2018-02-26 00:00:00                0.0  
...                             ...                ...  
99436           2017-03-28 00:00:00                1.0  
99437           2018-03-02 00:00:00                1.0  
99438           2017-09-27 00:00:00                1.0  
99439           2018-02-15 00:00:00                3.0  
99440           2018-04-03 00:00:00                1.0  

[99441 rows x 9 columns]

In [60]:
df_order['cancel_delay_days'].unique()

array([   2.,    1.,    0.,    3.,   nan,    5.,   12.,    4.,   18.,
         16.,    8.,    7.,    6.,   11.,   15.,   14.,   10.,    9.,
         13.,   -1.,   17.,   34.,   27.,   19.,   36.,   31.,   21.,
         29.,   22.,   23.,   20.,   30.,   28.,   38.,  104.,   24.,
         41.,   25.,   42.,   43.,   33.,   32.,   49.,   35.,   26.,
         40.,   44.,   47.,   39.,   37.,   46.,   45.,   55., -172.,
         54.,   53.,  107.,  125.,   52.,   66.,   48.,   62.,   -5.,
         61.])

In [61]:
inconsistent = df_order[
    (df_order['order_status'] == 'delivered') &
    (df_order['order_delivered_customer_date'].isna())
]

len(inconsistent)


8

In [62]:
df_order=df_order[~((df_order['order_status']=='delivered') & (df_order['order_delivered_customer_date'].isna()))]

In [63]:
df_order[(df_order['order_status']=='delivered') & (df_order['order_delivered_carrier_date'].isna())]

order_id                       customer_id  \
73222  2aa91108853cecb43c84a5dc5b277475  afeb16c7f46396c0ed54acb45ccaaa40   

      order_status order_purchase_timestamp   order_approved_at  \
73222    delivered      2017-09-29 08:52:58 2017-09-29 09:07:16   

      order_delivered_carrier_date order_delivered_customer_date  \
73222                          NaT           2017-11-20 19:44:47   

      order_estimated_delivery_date  cancel_delay_days  
73222           2017-11-14 00:00:00                NaN

In [64]:
df_order=df_order[~((df_order['order_status']=='delivered') & (df_order['order_delivered_carrier_date'].isna()))]

In [65]:
# To understand behavior:
def delivery_behavior(row):
     if pd.notna(row['order_delivered_customer_date']):
         return 'Delivered'
         
     if row['order_status'] == 'canceled' and pd.notna(row['order_delivered_carrier_date']):
         return 'Canceled_After_Shipping'
         
     if row['order_status'] == 'canceled':
        return 'Canceled_Before_Shipping'
     
     if row['order_status'] == 'shipped':
        return 'In_Transit'
         
     return 'Not_FulFilled'

df_order['Delivery_Behavior']=df_order.apply(delivery_behavior, axis=1)

In [66]:
df_order.head(20)

order_id                       customer_id  \
0   e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1   53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2   47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3   949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4   ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5   a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6   136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7   6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8   76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9   e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   
10  e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
11  34513ce0c4fab462a55830c0989c7edb  7711cf624183d843aafe81855097bc37   
12  82566a660a982b15fb86e904c8d32918  d3e3b74c766bc6214e0c830b17ee2341   
13  5ff96c15d0b717ac6ad1f3d77225a350  19402a48fe860416adf93348aba37740   
14  432aaf21d85167c2c86ec9448c4e42cc  3df704f53d3f1d4818840b34ec672a9f   
15  dcb36b511fcac050b97cd5c05de84dc3  3b6828a50ffe546942b7a473d70ac0fc   
16  403b97836b0c04a622354cf531062e5f  738b086814c6fcc74b8cc583f8516ee3   
17  116f0b09343b49556bbad5f35bee0cdf  3187789bec990987628d7a9beb4dd6ac   
18  85ce859fd6dc634de8d2f1e290444043  059f7fc5719c7da6cbafe370971a8d70   
19  83018ec114eee8641c97e08f7b4e926f  7f8c8b9c2ae27bf3300f670c3d478be8   

   order_status order_purchase_timestamp   order_approved_at  \
0     delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1     delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2     delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3     delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4     delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
5     delivered      2017-07-09 21:57:05 2017-07-09 22:10:13   
6      invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
7     delivered      2017-05-16 13:10:30 2017-05-16 13:22:11   
8     delivered      2017-01-23 18:29:09 2017-01-25 02:50:47   
9     delivered      2017-07-29 11:55:02 2017-07-29 12:05:32   
10    delivered      2017-05-16 19:41:10 2017-05-16 19:50:18   
11    delivered      2017-07-13 19:58:11 2017-07-13 20:10:08   
12    delivered      2018-06-07 10:06:19 2018-06-09 03:13:12   
13    delivered      2018-07-25 17:44:10 2018-07-25 17:55:14   
14    delivered      2018-03-01 14:14:28 2018-03-01 15:10:47   
15    delivered      2018-06-07 19:03:12 2018-06-12 23:31:02   
16    delivered      2018-01-02 19:00:43 2018-01-02 19:09:04   
17    delivered      2017-12-26 23:41:31 2017-12-26 23:50:22   
18    delivered      2017-11-21 00:03:41 2017-11-21 00:14:22   
19    delivered      2017-10-26 15:54:26 2017-10-26 16:08:14   

   order_delivered_carrier_date order_delivered_customer_date  \
0           2017-10-04 19:55:00           2017-10-10 21:25:13   
1           2018-07-26 14:31:00           2018-08-07 15:27:45   
2           2018-08-08 13:50:00           2018-08-17 18:06:29   
3           2017-11-22 13:39:59           2017-12-02 00:28:42   
4           2018-02-14 19:46:34           2018-02-16 18:17:02   
5           2017-07-11 14:58:04           2017-07-26 10:57:55   
6                           NaT                           NaT   
7           2017-05-22 10:07:46           2017-05-26 12:55:51   
8           2017-01-26 14:16:31           2017-02-02 14:08:10   
9           2017-08-10 19:45:24           2017-08-16 17:14:30   
10          2017-05-18 11:40:40           2017-05-29 11:18:31   
11          2017-07-14 18:43:29           2017-07-19 14:04:48   
12          2018-06-11 13:29:00           2018-06-19 12:05:52   
13          2018-07-26 13:16:00           2018-07-30 15:52:25   
14          2018-03-02 21:09:20           2018-03-12 23:36:26   
15          2018-06-11 14:54:00           2018-06-21 15:34:32   
16          2018-01-03 

In [67]:
df_order.drop(columns=['cancel_delay_days'],inplace=True)

In [68]:
df_order

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date Delivery_Behavior  
0               2017-10-18 00:00:00         Delivered  
1               2018-08-13 00:00:00         Delivered  
2               2018-09-04 00:00:00         Delivered  
3               2017-12-15 00:00:00         Delivered  
4               2018-02-26 00:00:00         Delivered  
...                             ...               ...  
99436           2017-03-28 00:00:00         Delivered  
99437           2018-03-02 00:00:00         Delivered  
99438           2017-09-27 00:00:00         Delivered  
99439           2018-02-15 00:00:00         Delivered  
99440           2018-04-03 00:00:00         Delivered  

[99432 rows x 9 columns]

In [69]:
df_order.drop(columns=['order_estimated_delivery_date'],inplace=True)

In [70]:
df_order

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      Delivery_Behavior  
0             Delivered  
1             Delivered  
2             Delivered  
3             Delivered  
4             Delivered  
...                 ...  
99436         Delivered  
99437         Delivered  
99438         Delivered  
99439         Delivered  
99440         Delivered  

[99432 rows x 8 columns]

In [71]:
df_order.to_csv("df_orders_cleaned.csv", index=False)

## Feature Engineering